cpp/src/neighbors/detail/vamana/vamana_build.cuh

```diff
@@ -0,0 +1,407 @@
+                                    raft::make_extents<int64_t>(max_batchsize, visited_size));
+
+  // Assign memory to query_list structures and initiailize
+  init_query_candidate_list<IdxT, accT><<<256, blockD, 0, stream>>>(query_list,
+                                                                    visited_ids.data_handle(),
+                                                                    visited_dists.data_handle(),
+                                                                    (int)max_batchsize,
+                                                                    visited_size);
+
+  // Create random permutation for order of node inserts into graph
+  std::vector<IdxT> insert_order;
+  create_insert_permutation<IdxT>(insert_order, (uint32_t)N);
+
+  // Memory needed to sort reverse edges - potentially large memory footprint
+  auto edge_dest =
+    raft::make_device_mdarray<IdxT>(res,
+                                    raft::resource::get_large_workspace_resource(res),
+                                    raft::make_extents<int64_t>(max_batchsize, degree));
+  auto edge_src =
+    raft::make_device_mdarray<IdxT>(res,
+                                    raft::resource::get_large_workspace_resource(res),
+                                    raft::make_extents<int64_t>(max_batchsize, degree));
+
+  size_t temp_storage_bytes = max_batchsize * degree * (2 * sizeof(IdxT));
+  RAFT_LOG_DEBUG("Temp storage needed for sorting (bytes): %lu", temp_storage_bytes);
+  auto temp_sort_storage =
+    raft::make_device_mdarray<IdxT>(res,
+                                    raft::resource::get_large_workspace_resource(res),
+                                    raft::make_extents<int64_t>(2 * max_batchsize, degree));
+
+  // Calculate the shared memory sizes of each kernel
+  int search_smem_sort_size = 0;
+  int prune_smem_sort_size  = 0;
+  SELECT_SMEM_SIZES(degree, visited_size);  // Sets above 2 variables to appropriate sizes
+
+  // Total dynamic shared memory used by GreedySearch
+  int search_smem_total_size =
+    static_cast<int>(search_smem_sort_size + dim * sizeof(T) + visited_size * sizeof(Node<accT>) +
+                     degree * sizeof(int) + queue_size * sizeof(DistPair<IdxT, accT>));
+
+  // Total dynamic shared memory size needed by both RobustPrune calls
+  int prune_smem_total_size =
+    prune_smem_sort_size + dim * sizeof(T) + (degree + visited_size) * sizeof(DistPair<IdxT, accT>);
+
+  RAFT_LOG_DEBUG("Dynamic shared memory usage (bytes): GreedySearch: %d, RobustPrune: %d",
+                 search_smem_total_size,
+                 prune_smem_total_size);
+
+  if (prune_smem_sort_size == 0) {  // If sizes not supported, smem sizes will be 0
+    RAFT_FAIL("Vamana graph parameters not supported: graph_degree=%d, visited_size:%d\n",
+              degree,
+              visited_size);
+  }
+
+  // Random medoid has minor impact on recall
+  int medoid_id = rand() % N;
```

Comment:
Yes it has been tested that a random medoid does not have any major impact. But DiskANN still uses another heuristic to select the medoid. Perhaps we want to mimic that. It is defined [here](https://github.com/microsoft/DiskANN/blob/main/src/in_mem_data_store.cpp#L341)

Reply:
I wanted to do this but ran out of time for this release. I created an issue to do this in the future: https://github.com/rapidsai/cuvs/issues/355

Reply:
Can you reference the issue in a todo comment in the code here? It helps because it adds a mapping between the issue and the code so we don't lose sight of it.

Reply:
Done.